### PR TITLE
consolidate binary_expr coercion rule code into `binary_rule.rs` module

### PR DESCRIPTION
**Pull request by alamb**
~Builds on https://github.com/apache/arrow-datafusion/pull/1606, so draft until that is merged~

# Which issue does this PR close?

Re https://github.com/apache/arrow-datafusion/issues/1605

 # Rationale for this change
While working on https://github.com/apache/arrow-datafusion/pull/1606 I found it confusing that some of the logic was in `coercion.rs` and some in `binary_rule.rs`

# What changes are included in this PR?
Moves all binary operator coercion logic into `binary_rule.rs` (just code motion, no changes now). I hope that the logic is consolidated more in a future PR

# Are there any user-facing changes?
no

### CODE REVIEW

**Comment by liukun4515** [comment]
> 

Thanks @alamb 
I think it's my fault and leave some duplicated code.
My original thought is that I will consolidate `binary logic and coercion` in the follow-up pull-requests.

**Comment by liukun4515** [comment]
If https://github.com/apache/arrow-datafusion/pull/1606 merged, I will review this.

**Comment by alamb** [comment]
> I think it's my fault and leave some duplicated code.
> My original thought is that I will consolidate binary logic and coercion in the follow-up pull-requests.

No worries @liukun4515  -- you said that. I just figured I was already changing the code for #1606  so I would make progress here too

File: datafusion/src/physical_plan/coercion_rule/binary_rule.rs [hunk 1]
```diff
@@ -15,15 +15,11 @@
 // specific language governing permissions and limitations
 // under the License.
 
-//! Support the coercion rule for binary operation
+//! Coercion rules for  matching argument types for binary operators
```
Comment (xudong963) [inline datafusion/src/physical_plan/coercion_rule/binary_rule.rs:18]: ocd, lol
```suggestion
//! Coercion rules for matching argument types for binary operators
```

**Review by xudong963 — APPROVED** [review]
LGTM, so much cleaner

File: datafusion/src/physical_plan/coercion_rule/binary_rule.rs [hunk 2]
```diff
@@ -293,14 +289,192 @@ fn coercion_decimal_mathematics_type(
     }
 }
 
+/// Determine if a DataType is signed numeric or not
+pub fn is_signed_numeric(dt: &DataType) -> bool {
```
Comment (liukun4515) [inline datafusion/src/physical_plan/coercion_rule/binary_rule.rs:293]: https://github.com/apache/arrow-datafusion/issues/1613

File: datafusion/src/physical_plan/coercion_rule/binary_rule.rs [hunk 3]
```diff
@@ -293,14 +289,192 @@ fn coercion_decimal_mathematics_type(
...
+            | DataType::Int64
+            | DataType::Float16
+            | DataType::Float32
+            | DataType::Float64
+            | DataType::Decimal(_, _)
+    )
+}
+
+/// Determine if a DataType is numeric or not
+pub fn is_numeric(dt: &DataType) -> bool {
```
Comment (liukun4515) [inline datafusion/src/physical_plan/coercion_rule/binary_rule.rs:308]: https://github.com/apache/arrow-datafusion/issues/1613

File: datafusion/src/physical_plan/coercion_rule/binary_rule.rs [hunk 4]
```diff
@@ -293,14 +289,192 @@ fn coercion_decimal_mathematics_type(
...
+                }
+            };
+
+            Some(Timestamp(unit, tz))
+        }
+        _ => None,
+    }
+}
+
+pub(crate) fn is_dictionary(t: &DataType) -> bool {
```
Comment (liukun4515) [inline datafusion/src/physical_plan/coercion_rule/binary_rule.rs:420]: https://github.com/apache/arrow-datafusion/issues/1613

**Review by liukun4515 — APPROVED** [review]
LGTM

**Pull request merged**